Client/service tutorial in rcl (#113)
micro-ROS/micro-ROS.github.io: _docs/tutorials/core/programming_rcl_rclc/index.md

---
title: Programming with rcl and rclc
permalink: /docs/tutorials/core/programming_rcl_rclc/
redirect_from: /docs/tutorials/core/
redirect_from: /docs/tutorials/
---

## Programming client/service with rcl

> ROS 2 services is another communication mechanism between nodes. Services implement a client-server paradigms based on ROS 2 messages and types. Further information about ROS 2 services can be found [here](https://index.ros.org/doc/ros2/Tutorials/Services/Understanding-ROS2-Services/)

Reviewer: ```suggestion
ROS 2 services is another communication mechanism between nodes. Services implement a client-server paradigm based on ROS 2 messages and types. Further information about ROS 2 services can be found [here](https://index.ros.org/doc/ros2/Tutorials/Services/Understanding-ROS2-Services/)
```

> Ready to use code related to this tutorial can be found in `micro-ROS-demos/rcl/addtwoints_server` and `micro-ROS-demos/rcl/addtwoints_client` folders.

Reviewer: I propose to link the names with the repo.

Reviewer 2: Done

> Starting from a code where RCL is init and a micro-ROS node is created, these steps are required in order to generate a service server:

Reviewer: ```suggestion
Starting from a code where RCL is initialized and a micro-ROS node is created, these steps are required in order to generate a service server:
```

```c
// Creating service server and options
rcl_service_options_t service_options = rcl_service_get_default_options();
rcl_service_t server = rcl_get_zero_initialized_service();

// Initializing service server
rcl_service_init(&server, &node, ROSIDL_GET_SRV_TYPE_SUPPORT(example_interfaces, srv, AddTwoInts), "addtwoints", &service_options);

// Init service server wait set
rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
rcl_wait_set_init(&wait_set, 0, 0, 0, 0, 1, 0, &context, rcl_get_default_allocator());

```

On the other hand the service client initialization looks like that:

```c
// Creating service client and options
rcl_client_options_t client_options = rcl_client_get_default_options();
rcl_client_t client = rcl_get_zero_initialized_client();

// Initializing service client
rcl_client_init(&client, &node, ROSIDL_GET_SRV_TYPE_SUPPORT(example_interfaces, srv, AddTwoInts), "addtwoints", &client_options)

// Init service client wait set
rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
rcl_wait_set_init(&wait_set, 0, 0, 0, 1, 0, 0, &context, rcl_get_default_allocator());
```

> First of all, seeing `AddTwoInts.srv` type definition it is possible to determine request and reply elements of the service. Service client will make a request with two integers and service server should send its sum as a response.

Author: ```suggestion
First of all, by looking at `AddTwoInts.srv` type definition it is possible to determine request and reply elements of the service. Service client will make a request with two integers and service server should send its sum as a response.
```

```
int64 a
int64 b
---
int64 sum
```

Once service client and server are configured, service client can perform a request and wait for reply:

```c
// Creating a service request
int64_t seq; 
example_interfaces__srv__AddTwoInts_Request req;
req.a = 24;
req.b = 42;

// Sending the request
rcl_send_request(&client, &req, &seq)
printf("Send service request %d + %d. Seq %ld\n",(int)req.a, (int)req.b, (int)seq);

// Wait for response
bool done = false;
do {
    rcl_wait_set_clear(&wait_set);

    size_t index;
    rcl_wait_set_add_client(&wait_set, &client, &index);

>     rcl_wait(&wait_set, RCL_MS_TO_NS(1));

Reviewer: For the first version, this can be left as is, but we should make use of the rclc executor soon. Please contact @JanStaschulat about the support for services in the executor.

Reviewer 2: We have filled the section of programming client/services on the rcl. Maybe @JanStaschulat can add a second subsections (or another section) showing an exampe of the rclc services API.
